Server.py

```diff
@@ -12,36 +12,43 @@
 serverSocket.listen(1) 
 list_of_connections = {}
 
+def user_login_server(username,password,conn): 
+    userObjList = structure.Users.values()
+    for userObj in userObjList: 
+        if (userObj.user_id == username) and (userObj.password == password):
+            conn.send("match")
+        else:
+            conn.send("nomatch")
 def add_user_server(user_id, new_user_id, chatroom_name, conn): 
+  
+  #hard-coded version 
   chatid = structure.Chatnames.get(chatroom_name)
+  currentUserObj = structure.Users.get(user_id)
   if chatid == None:
-    conn.send("Chatroom does not exist.") #is the 'b' a typo?
-  current_chat = structure.Chats.get(chatid) #Chat object
-  user_list = structure.demo_chat.chat_users
-  if new_user_id in structure.demo_chat.chat_users: #replace with current_chat later
-    conn.send("User is already in chat")
+    conn.send("Chatroom does not exist.".encode())  
+  this_chat_obj = structure.Chats.get(chatid)
+  
+  if new_user_id in this_chat_obj.chat_users:  
+    conn.send("User is already in chat".encode())
   else: 
-    #insert user id into current chat's user list
-    structure.demo_chat.chat_users.append(new_user_id) 
-    #can we store full names in the User object? Then I could have the user's name instead of username
+    this_chat_obj.chat_users.append(new_user_id) #how to add globally? 
     message = ("Added new user {uid} to the chat").format(uid = new_user_id)
-     #initialize chat's last_pushed_time to current time (for new user)
     structure.Chat_user3.last_pushed_time = datetime.datetime.now()  
```

Comment:
I wasn't super clear on this, but we want to update last_pushed_time in the user's Chat_user_obj within the Chatroom_obj. What's the best way to do this? I arbitrarily chose a Chat_user_obj to update (as a place to start), but would it be accessed via this_chat_obj.chat_users[-1].last_pushed_time = datetime.datetime.now()? 
this_chat_obj.chat_users[-1] --> the most recently added user